controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,477 @@
+			return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from ConfigMap: %v", err)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) reconcilePhases(clientMappingConfig *corev1.ConfigMap) (ctrl.Result, error) {
+	// Find the StorageClusterPeer from OwnerRef
+	owner := util.FindOwnerRefByKind(clientMappingConfig, "StorageClusterPeer")
+	if owner == nil {
+		return ctrl.Result{}, fmt.Errorf("failed to find StorgeClusterPeer owning the ClientMappingConfig")
+	}
+
+	// Fetch the StorageClusterPeer instance
+	storageClusterPeer := &ocsv1.StorageClusterPeer{}
+	storageClusterPeer.Name = owner.Name
+	storageClusterPeer.Namespace = clientMappingConfig.Namespace
+
+	if err := r.get(storageClusterPeer); err != nil {
+		if k8serrors.IsNotFound(err) {
+			r.log.Info("StorageClusterPeer resource not found. Ignoring since object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get StorageClusterPeer.")
+		return ctrl.Result{}, err
+	}
+
+	if storageClusterPeer.Status.State != ocsv1.StorageClusterPeerStatePeered {
+		return ctrl.Result{}, fmt.Errorf(
+			"waiting for StorageClusterPeer %s to be in Peered state",
+			storageClusterPeer.Name,
+		)
+	}
+
+	ocsClient, err := providerClient.NewProviderClient(r.ctx, storageClusterPeer.Spec.ApiEndpoint, util.OcsClientTimeout)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create a new provider client: %v", err)
+	}
+	defer ocsClient.Close()
+
+	if err = r.reconcileRbdMirror(clientMappingConfig); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if res, err := r.reconcileBlockPoolMirroring(
+		ocsClient,
+		clientMappingConfig,
+		storageClusterPeer,
+		true,
+	); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	if res, err := r.reconcileRadosNamespaceMirroring(
+		ocsClient,
+		clientMappingConfig,
+		storageClusterPeer); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) reconcileRbdMirror(clientMappingConfig *corev1.ConfigMap) error {
+	rbdMirror := &rookCephv1.CephRBDMirror{}
+	rbdMirror.Name = util.CephRBDMirrorName
+	rbdMirror.Namespace = clientMappingConfig.Namespace
+
+	storageConsumers := &ocsv1alpha1.StorageConsumerList{}
+	if err := r.list(
+		storageConsumers,
+		client.MatchingFields{util.AnnotationIndexName: util.RequestMaintenanceModeAnnotation},
+	); err != nil {
+		return fmt.Errorf("failed to list StorageConsumer(s): %v", err)
+	}
+
+	if len(storageConsumers.Items) > 1 {
+		if err := r.delete(rbdMirror); err != nil {
+			return fmt.Errorf("failed to delete CephRBDMirror: %v", err)
+		}
+	}
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, rbdMirror, func() error {
+		if err := r.own(clientMappingConfig, rbdMirror); err != nil {
+			return err
+		}
+		rbdMirror.Spec.Count = 1
+		return nil
+	})
+	if err != nil {
+		r.log.Error(err, "Failed to create/update the CephRBDMirror", "CephRBDMirror", rbdMirror)
+		return err
+	}
```

Comment:
1. Reverse the order, create first delete second. 
2. The condition for creating vs deleting should be more encompassing to include all known scenarios